src/v/datalake/translation/partition_translator.cc

```diff
@@ -0,0 +1,273 @@
+    // We factor in LSO to ensure only committed transactional batches are
+    // translated and we clamp it to committed offset to avoid translation
+    // of majorty appended but not flushed data (eg: acks=0/1/write_caching).
+    auto max_log_offset = std::min(
+      model::prev_offset(_partition->last_stable_offset()),
+      _partition->committed_offset());
+    return model::offset_cast(
+      _partition->get_offset_translator_state()->from_log_offset(
+        max_log_offset));
+}
+
+ss::future<std::optional<coordinator::translated_offset_range>>
+partition_translator::do_translation_for_range(
+  kafka::read_committed_reader rdr) {
+    // This configuration only writes a single row group per file but we limit
+    // the bytes via the reader max_bytes.
+    auto writer_factory = std::make_unique<batching_parquet_writer_factory>(
+      _writer_scratch_space,
+      fmt::format("rp-parquet-translator-term-{}", _term),
+      max_rows_per_row_group,
+      max_bytes_per_row_group);
+    auto multiplexer = datalake::record_multiplexer(std::move(writer_factory));
+    auto result = co_await std::move(rdr).consume(
+      std::move(multiplexer), model::no_timeout);
+    if (result.has_error()) {
+        vlog(_logger.warn, "Error translating range {}", result.error());
+        co_return std::nullopt;
+    }
+    co_return std::move(result.value());
+}
+
+ss::future<partition_translator::translation_success>
+partition_translator::do_translate_once() {
+    if (
+      !_partition->get_ntp_config().iceberg_enabled()
+      || !_features->local().is_active(features::feature::datalake_iceberg)) {
+        co_return translation_success::no;
+    }
+    auto reconcile_result = co_await reconcile_with_coordinator();
+    if (!reconcile_result) {
+        vlog(_logger.warn, "reconciliation with coordinator failed");
+        co_return translation_success::no;
+    }
+    vlog(_logger.trace, "reconciliation result: {}", reconcile_result.value());
+    auto read_begin_offset = kafka::next_offset(reconcile_result.value());
+    auto read_end_offset = max_offset_for_translation();
+    if (read_begin_offset >= read_end_offset) {
+        vlog(
+          _logger.trace,
+          "translation up to date until kafka offset: {}, lso: {}, nothing to "
+          "do",
+          _partition->last_stable_offset());
+        co_return translation_success::yes;
+    }
+    // We have some data to translate, make a reader
+    // and dispatch to the iceberg translator
+    //
+    // todo: what if we have too little to read, eg: slow moving topic.
+    // Currently we rely on user setting the translation interval
+    // appropriately to ensure each attempt has something substantial
+    // to read. We can add some heuristics to estimate the size and
+    // backoff if it is too small and wait for enough data to
+    // accumulate. The resulting parquet files are only performant
+    // if there is a big chunk of data in them. Smaller parquet files
```

Comment:
I think here we can use `disk_log_impl::size_bytes_after_offset` to first query the possible read size, this way we can know for sure what the read size is. I think this is required for v1 as without that we may end up with lots of tiny parque files.

Reply:
right, I did consider `disk_io_impl::offset_range_size()` which seems like a more accurate version of size_bytes_after_offset(), left it out for v1, we still have to give up at some point right?  

eg: there is only little data to read but it has been long time since since last translation at which point we should probably just translate even if the data is small.

Reply:
chatted offline: this heuristic will be in a separate patch.